### ISSUE TITLE
OmniAuth possibly broken with EditingAuth

### ISSUE THREAD

**Issue by explody** [gollum/gollum#940]
Our running wiki has been using omnigollum w/ SAML & LDAP auth for quite some time.  After upgrade, auth was failing, AFAICT due to this:  https://github.com/arr2036/omnigollum/issues/24

I explicitly set allow_editing to true but then ran into a weird error and finally dug out this exception:

```
NoMethodError at /__omnigollum__/auth/saml/callback
private method `settings' called for #<OmniAuth::Builder:0x007f58be5e5740>

/srv/wikiops/gollum/lib/gollum/editing_auth.rb in call
    def initialize(app)
      @app = app
    end

    def call(env)
      @env = env
      # Blocks all potentially editable pages. Use EditingAuth::whitelist_pages to unblock pages.
      unless (env["REQUEST_METHOD"] == "GET") || @app.settings.wiki_options[:allow_editing]
        return block unless excluded_page?
      end
      @app.call(env)
    end

    def block
      [403, {'Content-Type' => 'text/html', 'Content-Length' => '9'}, ['Forbidden']]

```

Obviously, this well may be an omniauth/omnigollum issue (I really can't tell for sure).  What I can tell is that the exception hits inside editing_auth.rb when it tries to access @app.settings.  This seems completely reasonable but only showed up after the addition of EditingAuth.  If you think this should go to omnigollum/omniauth, I'll repost over there.

I worked around this by editing app.rb and completely removing EditingAuth for the moment.

Thanks #2 ;)

**Comment by dometto** [comment]
Thanks for reporting this! I'm not familiar at all with the internal workings of omniauth/omnigollum, but as you said it should be reasonable for gollum to access `@app.settings` here. I'm a bit surprised that `@app` is apparently an instance of `OmniAuth::Builder` rather than an instance of `Precious::App` (but that may reflect my omniauth ignorance), which makes me think that this might indeed be an omnigollum issue. Could you open an issue there? I'll be happy to work with the omnigollum folks if that's necessary to fix this.

**Comment by explody** [comment]
Agreed all around.  Definitely looks like omniauth is intercepting ```app``` which doesn't seem entirely weird, but why ```app.settings``` would suddenly be private, I don't know.  I'll bounce this over to omnigollum and see what happens.

**Comment by bartkamphorst** [comment]
@explody Any news from the good folks over at omnigollum? Wondering if I can close this issue here.

**Comment by ringe** [comment]
@bartkamphorst this issue is very much present. This prevents any POST requests on my wiki.

**Comment by dometto** [comment]
@ringe, our hunch was that this is an issue with omnigollum. Any idea what might be causing this?

**Comment by ringe** [comment]
Here's a clue. I wanted to see what object ```@app``` is in editing_auth.rb:

```
    def initialize(app)
      @app = app
    end

    def call(env)
      @env = env
      puts "\n\n"+@app.to_s+"\n\n"
      # Blocks all potentially editable pages. Use EditingAuth::whitelist_pages to unblock pages.
      unless (env["REQUEST_METHOD"] == "GET") || @app.settings.wiki_options[:allow_editing]
        return block unless excluded_page?
      end
      @app.call(env)
    end
```
This is the ningx error.log output:
```
App 24804 stdout: #<Sinatra::Wrapper:0x007f1595534010>
App 24804 stdout: 
App 24804 stdout: 
App 24804 stdout: 
App 24804 stdout: 
App 24804 stdout: #<Precious::EditingAuth:0x007f1595be5c60>
App 24804 stdout: 
App 24804 stderr: [ 2015-02-10 12:01:02.6803 24804/0x007f1595e66368(Worker 1) utils.rb:84 ]: *** Exception NoMethodError in Rack application object (private method `settings' called for #<OmniAuth::Builder:0x007f1595be6818>) (process 24804, thread 0x007f1595e66368(Worker 1)):
App 24804 stderr: 	from /home/deploy/doc/vendor/bundle/ruby/2.1.0/gems/gollum-3.1.2/lib/gollum/editing_auth.rb:10:in `call'
App 24804 stderr: 	from /home/deploy/doc/vendor/bundle/ruby/2.1.0/gems/rack-protection-1.5.3/lib/rack/protection/xss_header.rb:19:in `call'
```

So I tried to remove the initialize method, because I notice that the xss_header.rb does not seem to need one.

It looks to me like Rack overrides the ```@app ``` variable between the initializer and the call methods, since this is the output after removing the initializer method:
```
App 25947 stdout: #<OmniAuth::Builder:0x007fb2be539210>
App 25947 stdout: 
App 25947 stderr: [ 2015-02-10 12:06:19.9057 25947/0x007fb2be786ae0(Worker 1) utils.rb:84 ]: *** Exception NoMethodError in Rack application object (private method `settings' called for #<OmniAuth::Builder:0x007fb2be539210>) (process 25947, thread 0x007fb2be786ae0(Worker 1)):
App 25947 stderr: 	from /home/deploy/doc/vendor/bundle/ruby/2.1.0/gems/gollum-3.1.2/lib/gollum/editing_auth.rb:7:in `call'
App 25947 stderr: 	from /home/deploy/doc/vendor/bundle/ruby/2.1.0/gems/rack-protection-1.5.3/lib/rack/protection/xss_header.rb:18:in `call'
```

So the problem is that settings is called on Omniauth::Builder, but I will try and locate where ```@app``` is set to Omniauth::Builder rather than Precious::App.

**Comment by dometto** [comment]
> I will try and locate where @app is set to Omniauth::Builder rather than Precious::App.

:+1: Thanks!

**Comment by aldanor** [comment]
Any updates on this? :)

**Comment by bartkamphorst** [comment]
@ringe Any luck locating where `@app` is set to Omniauth::Builder?

**Comment by tanzim** [comment]
@explody @ringe Any update? I can't upgrade from Gollum 3.1.0 till this is fixed. I'd love to help, but my Ruby is very basic!

**Comment by dometto** [comment]
Done some digging, posted this in the omnigollum issue:

> The core problem is caused by the presence of the new `EditingAuth` middleware in gollum (see [here](https://github.com/gollum/gollum/issues/940)). That middleware is expecting the app it is passed to be Gollum, but when using omnigollum, it's in fact handed an instance of `OmniAuth::Builder`. That's done by omnigollum [here](https://github.com/arr2036/omnigollum/blob/master/lib/omnigollum.rb#L217). 

> It does work fine when I add the `EditingAuth` middleware *after* omnigollum puts `OmniAuth::Builder` on the middleware stack. However, to do so, I need to do `Precious::App.use Precious::EditingAuth` outside of `Precious:App`. That won't work when using gollum as a rack app (e.g. with `config.ru` and `rackup`), because then `EditingAuth` won't get used at all unless the user manually adds it.

> Bottom line: is there a way for omnigollum to ensure that `OmniAuth::Builder` is put on the middleware stack in front of the other middleware? I don't see a way to change the ordering of middleware in Sinatra. Sinatra's `use` just appends the middleware to the stack.

@explody @ringe, any help is much appreciated. :)

**Comment by dometto** [comment]
Re-implemented the blocking of non-GET requests without the piece of middleware. That resolves the problem for me locally. @explody @ringe @aldanor: can you confirm that the problem dissappears on gollum `HEAD`? If so I'll quickly do a bugfix release.

**Comment by dometto** [comment]
Released in 3.1.3
